kips/kip-2.md

```diff
@@ -0,0 +1,462 @@
+    require(!pendingAdmins[_addr], 'already proposed as an admin');
+    pendingAdmins[_addr] = true;
+    emit AdminProposed(_addr);
+  }
+
+  function cancelNewAdminProposal(address _addr) public assertIsAdmin {
+    require(pendingAdmins[_addr], 'not proposed as an admin');
+    pendingAdmins[_addr] = false;
+    emit AdminProposalCancelled(_addr);
+  }
+
+  function acceptAdminRole() public {
+    require(pendingAdmins[msg.sender], 'not proposed as an admin');
+    pendingAdmins[msg.sender] = false;
+    admins[msg.sender] = true;
+    numAdmins++;
+    emit AdminProposalAccepted(msg.sender);
+  }
+
+  function removeAdmin(address _addr) public assertIsAdmin {
+    require(1 < numAdmins, 'cannot remove last admin');
+    require(_addr != msg.sender, 'cannot remove oneself');
+    require(admins[_addr], 'not an admin');
+    admins[_addr] = false;
+    numAdmins--;
+    emit AdminRemoved(_addr);
+  }
+}
+
+/**
+ * Access control helper base class.
+ */
+contract AccessControl is EternalStorage {
+  modifier assertIsAdmin () {
+    require(acl().isAdmin(msg.sender), 'must be admin');
+    _;
+  }
+
+  constructor (address _acl) public {
+    dataAddress["acl"] = _acl;
+  }
+
+  function acl () internal view returns (IACL) {
+    return IACL(dataAddress["acl"]);
+  }
+}
+```
+
+**UserPot, IUserPot, UserPotImpl**
+
+The user pot.
+
+Note that `UserPot` doesn't implement `IUserPot` itself since it's going to be a proxy contract which delegates all calls to a `UserPotImpl` instance. Also note that `upgrade()` can only be called by an administrator.
+
+```solidity
+/**
+ * Interface for User pot
+ */
+interface IUserPot {
+  function deposit(address _event, address _user, uint256 _deposit) external payable;
+}
+
+
+/**
+ * User pot instance
+ */
+contract UserPot is AccessControl, Proxy {
+  constructor (address _acl, address _userPotImpl) AccessControl(_acl) Proxy(_userPotImpl) public {}
+
+  function upgrade (address _implementation) public assertIsAdmin {
+    setImplementation(_implementation);
+  }
+}
+
+/**
+ * User pot implementation
+ */
+contract UserPotImpl is EternalStorage, AccessControl, IUserPot {
+  function deposit(address _event, address _user, uint256   _deposit) public payable {
+    IEvent _event = IEvent(msg.sender);
+    uint256 _deposit = _event.getDeposit(_user);
+    uint256 bal = calculatePayout(_user);
+    bal += msg.value;
+    require(bal >= _deposit, 'you need to pay more to register for event');
+    _updateUserData(_user, msg.sender, bal - _deposit);          
+  }
+
+  function withdraw() public {
+    uint256 bal = calculatePayout(msg.sender);
+    _updateUserData(msg.sender, address(0), 0);
+    msg.sender.transfer(bal);
+  }
+
+  function calculatePayout(address _user) public view returns (uint256) {
+    string memory key = string(abi.encodePacked(_user));
+    uint256 bal = dataUint256[key];
+    address[] storage events = dataManyAddresses[key];
+    for (uint256 i = 0; i < events.length; i += 1) {
```

Comment:
What would be the incremental gas cost of running this as users go to more event? Assuming event data are coming from eternal storage, can you even upgrade?